File: accommodation/accommodation/report/hotel_trends/hotel_trends.js
// Copyright (c) 2016, DBF and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Hotel Trends"] = {
	"filters": [
		{
			"fieldname": "report",
			"label": __("Report"),
			"fieldtype": "Select",
			"options": ["Room Type Occupancy", "Room Occupancy"].join('\n'),
			"default": "Room Occupancy",
			"reqd": 1
		},
		{
			"fieldname": "check_in",
			"label": __("From Date"),
			"fieldtype": "Date",
			"default": frappe.datetime.add_days(frappe.datetime.now_date(), -7),
			"reqd": 1
		},
		{
			"fieldname": "check_out",
			"label": __("To Date"),
			"fieldtype": "Date",
			"default": frappe.datetime.now_date(),
			"reqd": 1
		},
	],
	
	"formatter": function(value, row, column, data, default_formatter) {

		if (["room_type","room"].indexOf(column.fieldname) < 0) {
			if (value != 0) {
				value = $(`<span>${value}</span>`);
				var $value = $(value).css({
					"margin": "0px",
					"padding-left":"5px",
					"padding-right":"2px",
					"background-color": "#ffcf99",
					"font-weight":"bold",
					"color": "red",
					"width": "auto",
					"display": "block"
				});
			} else {
				value = $(`<span>${value}</span>`);
				var $value = $(value).css({
					"margin": "0px",
					"padding-left":"5px",
					"padding-right":"2px",
					"font-weight":"bold",
					"color": "green",
					"width": "auto",
					"display": "block"
				});
			}
			// "background-color": "#ffcf99",

			value = $value.wrap("<p></p>").parent().html();
		}
		return value;
	}
}

File: accommodation/accommodation/report/hotel_trends/hotel_trends.py
# Copyright (c) 2013, DBF and contributors
# For license information, please see license.txt

from __future__ import unicode_literals
import frappe

def execute(filters=None):
    report = filters.get("report")
    if report == "Room Type Occupancy":
        columns, data = get_room_type_occupancy(filters)
    if report == "Room Occupancy":
        columns, data = get_room_occupancy(filters)

    return columns, data


def get_room_type_occupancy(filters):

    query = """
select
    d.db_date, room.room_type, coalesce(sum(if(r.room is null,0,1)),0) occ
from 
    tabDate d
cross join 
    `tabAccommodation Room` room 
left outer join 
    `tabAccommodation Booking Item` r on d.db_date between r.check_in and r.check_out  and room.name = r.room
where 
    d.db_date between %(check_in)s and %(check_out)s
    and r.item_status <> 'Cancelled'
group by 
    d.db_date, room.room_type
"""
    data = frappe.db.sql(query, filters, as_list=1)

    import pandas as pd
    df = pd.DataFrame(data, columns=['date', 'room_type', 'occ'])
    pivot = pd.pivot_table(df, index=['room_type'], columns=[
        'date'], values='occ', aggfunc=sum, margins=True, margins_name='Total', fill_value=0, dropna=True)

    from accommodation.accommodation import pivot_to_report
    return pivot_to_report(pivot)

def get_room_occupancy(filters):

    query = """
select
    d.db_date, room.room_type, room.name, coalesce(sum(if(r.room is null,0,1))) occ
from 
    tabDate d
cross join 
    `tabAccommodation Room` room 
left outer join 
    `tabAccommodation Booking Item` r on d.db_date between r.check_in and r.check_out  and room.name = r.room
where 
    d.db_date between %(check_in)s and %(check_out)s
    and r.item_status <> 'Cancelled'
group by 
    d.db_date, room.room_type, room.name
"""
    data = frappe.db.sql(query, filters, as_list=1)

    import pandas as pd
    df = pd.DataFrame(data, columns=['date', 'room_type', 'room', 'occ'])
    pivot = pd.pivot_table(df, index=['room_type', 'room'], columns=[
        'date'], values='occ', aggfunc=sum, margins=True, margins_name='Total', fill_value=0, dropna=True)

    from accommodation.accommodation import pivot_to_report
    return pivot_to_report(pivot)